src/uipath/runtime/base.py

```diff
@@ -154,101 +99,101 @@ async def stream(
         # Without it, the function wouldn't match the AsyncGenerator return type
         yield
 
-    @abstractmethod
-    async def validate(self):
-        """Validate runtime inputs."""
-        pass
+    def get_instrumentor(self) -> Optional[BaseInstrumentor]:
+        """Get instrumentor for this runtime. If no instrumentor is available, return None."""
+        return None
 
     @abstractmethod
     async def cleanup(self):
         """Cleaup runtime resources."""
         pass
 
-    async def __aexit__(self, exc_type, exc_val, exc_tb):
-        """Async exit method called when exiting the 'async with' block.
 
-        Cleans up resources and handles any exceptions.
+T = TypeVar("T", bound=UiPathBaseRuntime)
 
-        Always writes output file regardless of whether execution was successful,
-        suspended, or encountered an error.
-        """
+
+class UiPathExecutionRuntime(UiPathBaseRuntime, Generic[T]):
+    """Handles runtime execution with tracing/telemetry."""
+
+    def __init__(
+        self,
+        delegate: T,
+        trace_manager: UiPathTraceManager,
+        root_span: str = "root",
+        execution_id: Optional[str] = None,
+    ):
+        """Initialize the executor."""
+        self.delegate = delegate
+        self.trace_manager = trace_manager
+        self.root_span = root_span
+        self.execution_id = execution_id
+        self.log_handler: Optional[UiPathRuntimeExecutionLogHandler]
+        if execution_id is not None:
+            self.log_handler = UiPathRuntimeExecutionLogHandler(execution_id)
+        else:
+            self.log_handler = None
+
+    async def execute(
+        self,
+        input: dict[str, Any],
+    ) -> Any:
+        """Execute runtime with context."""
+        instrumentor = self.delegate.get_instrumentor()
+        log_interceptor = UiPathRuntimeLogsInterceptor(
+            execution_id=self.execution_id, log_handler=self.log_handler
+        )
+        log_interceptor.setup()
+
+        if instrumentor is not None:
+            instrumentor.instrument(tracer_provider=self.trace_manager.tracer_provider)
         try:
-            if self.context.result is None:
-                execution_result = UiPathRuntimeResult()
+            if self.execution_id:
+                with self.trace_manager.start_execution_span(
+                    self.root_span, execution_id=self.execution_id
+                ):
+                    return await self.delegate.execute(input)
             else:
-                execution_result = self.context.result
```

Comment:
this piece of code with `aenter` and `aexit` is very important (it represents the contract with the serverless runtime). moved it to the `UiPathRuntimeContext`. Now we can do something like this in the CLI and make sure the logs and errors are properly captured and reported.
```python
with UiPathRuntimeContext.with_defaults(...) as ctx:
     ctx.result = await runtime.execute(input)
```